flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorPartitionLifecycleTest.java

```diff
@@ -102,6 +126,115 @@ public void setup() {
 		rpc = new TestingRpcService();
 	}
 
+	@Test
+	public void testConnectionTerminationAfterExternalRelease() throws IOException, InterruptedException, ExecutionException, TimeoutException {
+		final JobMasterId jobMasterId = JobMasterId.generate();
+
+		final LibraryCacheManager libraryCacheManager = mock(LibraryCacheManager.class);
+		when(libraryCacheManager.getClassLoader(any(JobID.class))).thenReturn(ClassLoader.getSystemClassLoader());
+
+		final JobMasterGateway jobMasterGateway = mock(JobMasterGateway.class);
+		when(jobMasterGateway.getFencingToken()).thenReturn(jobMasterId);
+
+		final TaskManagerActions taskManagerActions = new NoOpTaskManagerActions();
+		final JobManagerConnection jobManagerConnection = new JobManagerConnection(
+			jobId,
+			ResourceID.generate(),
+			jobMasterGateway,
+			taskManagerActions,
+			mock(CheckpointResponder.class),
+			new TestGlobalAggregateManager(),
+			libraryCacheManager,
+			new NoOpResultPartitionConsumableNotifier(),
+			mock(PartitionProducerStateChecker.class));
+
+		final JobManagerTable jobManagerTable = new JobManagerTable();
+		jobManagerTable.put(jobId, jobManagerConnection);
+
+		final AtomicBoolean hasPartitionsOccupyingLocalResources = new AtomicBoolean(true);
+		final TestJobAwareShuffleEnvironment jobAwareShuffleEnvironment = new TestJobAwareShuffleEnvironment(jobId -> hasPartitionsOccupyingLocalResources.get());
+
+		final TaskManagerServices taskManagerServices = new TaskManagerServicesBuilder()
+			.setJobManagerTable(jobManagerTable)
+			.setShuffleEnvironment(jobAwareShuffleEnvironment)
+			.build();
+
+		final TestingTaskExecutor taskManager = createTestingTaskExecutor(taskManagerServices, new HeartbeatServices(Long.MAX_VALUE, Long.MAX_VALUE));
+
+		try {
+			taskManager.start();
+			taskManager.waitUntilStarted();
+
+			assertTrue(taskManagerServices.getJobManagerTable().contains(jobId));
+
+			taskManager.releasePartitions(jobId, Collections.singletonList(new ResultPartitionID()));
+			// connection should be kept alive since the environment still says we have local resources
+			assertTrue(taskManagerServices.getJobManagerTable().contains(jobId));
+
+			hasPartitionsOccupyingLocalResources.set(false);
+
+			taskManager.releasePartitions(jobId, Collections.singletonList(new ResultPartitionID()));
+			assertTrue(taskManagerServices.getJobManagerTable().contains(jobId));
```

Comment:
We should expect `assertFalse` here after solving the NPE issue caused by mocked `TaskSlotTable` mentioned above.